### PR TITLE
image: add registry-url flag for verify-image-signature command

### PR DESCRIPTION
**Pull request by mfojtik**
If the registry is exposed via route and the `oc` client does not have access to internal registry, we need to provide a hint to the verification to be able to fetch the image manifest. The registry-url overrides the image registry URL and it allows verification of images outside the cluster.

### CODE REVIEW

**Comment by mfojtik** [comment]
@miminar @thomasmckay FYI

File: pkg/oc/admin/image/verify-signature.go [hunk 1]
```diff
@@ -108,6 +119,8 @@ func NewCmdVerifyImageSignature(name, fullName string, f *clientcmd.Factory, out
 	cmd.Flags().BoolVar(&opts.Save, "save", opts.Save, "If true, the result of the verification will be saved to an image object.")
 	cmd.Flags().BoolVar(&opts.RemoveAll, "remove-all", opts.RemoveAll, "If set, all signature verifications will be removed from the given image.")
 	cmd.Flags().StringVar(&opts.PublicKeyFilename, "public-key", opts.PublicKeyFilename, fmt.Sprintf("A path to a public GPG key to be used for verification. (defaults to %q)", opts.PublicKeyFilename))
+	cmd.Flags().StringVar(&opts.RegistryURL, "registry-url", opts.RegistryURL, fmt.Sprintf("The address to use when contacting the registry, instead of using the internal cluster address. This is useful if you can't resolve or reach the internal registry address.", opts.PublicKeyFilename))
```
Comment (miminar) [inline pkg/oc/admin/image/verify-signature.go]: Superfluous `, opts.PublicKeyFilename`

File: pkg/oc/admin/image/verify-signature.go [hunk 2]
```diff
@@ -108,6 +119,8 @@ func NewCmdVerifyImageSignature(name, fullName string, f *clientcmd.Factory, out
...
+	cmd.Flags().BoolVar(&opts.Insecure, "insecure", opts.Insecure, "If set, use the insecure (http) protocol for the registry-url parameter.")
```
Comment (miminar) [inline pkg/oc/admin/image/verify-signature.go]: I believe the insecure flag should be honored regardless of internal or external route.

File: pkg/oc/admin/image/verify-signature.go [hunk 3]
```diff
@@ -143,8 +159,6 @@ func (o *VerifyImageSignatureOptions) Complete(f *clientcmd.Factory, cmd *cobra.
 
 	// We need the current user name so we can record it into an verification condition and
 	// we need a bearer token so we can fetch the manifest from the registry.
-	// TODO: Add support for external registries (currently only integrated registry will
```
Comment (miminar) [inline pkg/oc/admin/image/verify-signature.go]: In this context, the *external* means the hosted outside of cluster, right? As such, this looks still valid.

File: pkg/oc/admin/image/verify-signature.go [hunk 4]
```diff
@@ -108,6 +119,8 @@ func NewCmdVerifyImageSignature(name, fullName string, f *clientcmd.Factory, out
 	cmd.Flags().BoolVar(&opts.Save, "save", opts.Save, "If true, the result of the verification will be saved to an image object.")
 	cmd.Flags().BoolVar(&opts.RemoveAll, "remove-all", opts.RemoveAll, "If set, all signature verifications will be removed from the given image.")
 	cmd.Flags().StringVar(&opts.PublicKeyFilename, "public-key", opts.PublicKeyFilename, fmt.Sprintf("A path to a public GPG key to be used for verification. (defaults to %q)", opts.PublicKeyFilename))
+	cmd.Flags().StringVar(&opts.RegistryURL, "registry-url", opts.RegistryURL, "The address to use when contacting the registry, instead of using the internal cluster address. This is useful if you can't resolve or reach the internal registry address.")
+	cmd.Flags().BoolVar(&opts.Insecure, "insecure", opts.Insecure, "If set, use the insecure (http) protocol for the registry-url parameter.")
```
Comment (miminar) [inline pkg/oc/admin/image/verify-signature.go]: The insecure flag should be honored regardless of internal or external route.

Comment (mfojtik) [reply pkg/oc/admin/image/verify-signature.go]: will check how I can plumb it for internal route

**Comment by mfojtik** [comment]
@miminar updated

@thomasmckay with the latest change, you will have to pass the `--insecure` parameter during verification (default is secure)

File: pkg/oc/admin/image/verify-signature.go [hunk 5]
```diff
@@ -44,6 +45,9 @@ var (
 	key or invalid expected identity will cause the saved verification status to be removed
 	and the image will become "unverified".
 
+	If this command is ran outside the cluster, users have to specify the "--registry-url" parameter
```
Comment (bparees) [inline pkg/oc/admin/image/verify-signature.go]: s/ran/run/

File: pkg/oc/admin/image/verify-signature.go [hunk 6]
```diff
@@ -108,6 +119,8 @@ func NewCmdVerifyImageSignature(name, fullName string, f *clientcmd.Factory, out
 	cmd.Flags().BoolVar(&opts.Save, "save", opts.Save, "If true, the result of the verification will be saved to an image object.")
 	cmd.Flags().BoolVar(&opts.RemoveAll, "remove-all", opts.RemoveAll, "If set, all signature verifications will be removed from the given image.")
 	cmd.Flags().StringVar(&opts.PublicKeyFilename, "public-key", opts.PublicKeyFilename, fmt.Sprintf("A path to a public GPG key to be used for verification. (defaults to %q)", opts.PublicKeyFilename))
+	cmd.Flags().StringVar(&opts.RegistryURL, "registry-url", opts.RegistryURL, "The address to use when contacting the registry, instead of using the internal cluster address. This is useful if you can't resolve or reach the internal registry address.")
```
Comment (bparees) [inline pkg/oc/admin/image/verify-signature.go]: also useful if you just want to verify the signature of an image on some other registry other than the cluster registry?

File: pkg/oc/admin/image/verify-signature.go [hunk 7]
```diff
@@ -123,6 +136,9 @@ func (o *VerifyImageSignatureOptions) Validate() error {
 	if o.RemoveAll && len(o.ExpectedIdentity) > 0 {
 		return errors.New("the --expected-identity cannot be used when removing all verifications")
 	}
+	if len(o.RegistryURL) == 0 && o.Insecure {
+		return errors.New("the --insecure can be used only with --registry-url")
```
Comment (bparees) [inline pkg/oc/admin/image/verify-signature.go]: s/--insecure can be used only with/--insecure flag can only be used with/

File: pkg/oc/admin/image/verify-signature.go (hunk 6 above)
Comment (mfojtik) [reply pkg/oc/admin/image/verify-signature.go]: in theory yes, in practice this need openshift api to deal with signatures and registry just to fetch the manifest ;-)

File: pkg/oc/admin/image/verify-signature.go (hunk 7 above)
Comment (miminar) [reply pkg/oc/admin/image/verify-signature.go]: I believe the whole block can be deleted now. The flag should be honored for both routes.

Comment (mfojtik) [reply pkg/oc/admin/image/verify-signature.go]: yeah, this is not needed anymore

**Comment by mfojtik** [comment]
comments addressed, re-PTAL ;-)

**Comment by mfojtik** [comment]
/retest

(etcd timeout ...)

**Pull request merged**